### PR TITLE
basic implementation of contextmenu

### PR DESCRIPTION
**Pull request by Korede-TA**
this mostly works, except the context menu is rendered a number of pixels below where it actually should be. this can probably be fixed with a quick hack but i'll try to find something more lasting first.

### CODE REVIEW

File: src/view.rs [hunk 1]
```diff
@@ -717,3 +716,43 @@ pub fn view_grid_grammar(m: &Model, coord: &Coordinate, sub_coords: Vec<Coordina
         </div>
     }
 }
+
+pub fn view_context_menu(m: &Model) -> Html {
```
Comment (amad1101) [inline src/view.rs]: perhaps a better name instead of using for readability.

**Comment by SirD-AL** [comment]
- Corrected positioning of the context menu

- Added the closing click that still needs a bit of code to work everywhere correctly

- Added the switch template for Exceptions, that I will enhance with what we have and is opened for future functionalities.

**EXAMPLE**: Merge in the next screenshots

Screen Shot 2020-04-09 at 11 28 57 AM
Screen Shot 2020-04-09 at 11 29 11 AM

File: src/view.rs (hunk 1 above)
Comment (Korede-TA) [reply src/view.rs:720]: better name such as...? i think this one is pretty explanatory.

File: src/model.rs [hunk 2]
```diff
@@ -1168,6 +1171,7 @@ impl Component for Model {
                         // context menu
                         oncontextmenu=self.link.callback(move |e: ContextMenuEvent| {
                             e.prevent_default();
+                            info!("X: {:?}", e.client_x());
```
Comment (Korede-TA) [inline src/model.rs:1174]: remove log statement if it's not necessary

File: src/view.rs [hunk 3]
```diff
@@ -719,25 +719,43 @@ pub fn view_grid_grammar(m: &Model, coord: &Coordinate, sub_coords: Vec<Coordina
...
-        for (option_name, option_action) in default_options {
-            v.add_child(html! {
-                <li class="context-menu-option" onclick=option_action>
-                    { option_name }
-                </li>
-            });
+        for (option_name, option_action, option_param) in default_options {
+            let mut param = true;
+            
+            //Switch exceptions
```
Comment (Korede-TA) [inline src/view.rs:738]: could you add a longer explanation of what you're doing here? i don't quite understand by just looking at it.

Comment (Korede-TA) [reply src/view.rs:738]: it seems like these refer to options that should only show based on conditions like the selection. try to make that as obvious as possible in the code by using more obvious variable names, or adding more comments

**Comment by SirD-AL** [comment]
1. gif for single cell open/closing
SingleGif

2. gif for double cell open/close
Double Gif

**Don't Merge Yet**

**Comment by SirD-AL** [comment]
- New layering added.

- Layering conditional is on the work and new arguments have been added :

> (option_name, option_action, option_param, option_layer)

Example
`("Zoom In (+)", m.link.callback(|_| Action::ZoomIn), true, 2),`

> option Name and action are what their name means
    option_param represents the default or conditionnal render of an option
    option_layer represents the visual layer of the option on the context menu that for now only helps the break however will evolve in the future

Design will evolve and conditions are open and modifiable with all the functionalities we will add later.

image

**Comment by Korede-TA** [comment]
@SirD-AL this looks really good. separating the options visually like that is worthwhile. are you gonna do the style changes in this PR as well?

otherwise, lemme look at the code and then i'll approve.

**Pull request merged**